crypto/bio/bio_lib.c

```diff
@@ -786,41 +786,48 @@ void bio_cleanup(void)
 }
 
 /* Internal variant of the below BIO_wait() not calling BIOerr() */
-static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+static int bio_wait(BIO *bio, time_t max_time, unsigned int polling_millisec)
 {
 #ifndef OPENSSL_NO_SOCK
     int fd;
 #endif
+    long sec_diff;
 
-    if (max_time == 0)
+    if (max_time == 0) /* no timeout */
         return 1;
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
         return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
 #endif
-    if (milliseconds > 1000) {
-        long sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    /* fall back to polling since no sockets are available */
 
-        if (sec_diff <= 0)
-            return 0; /* timeout */
-        if ((unsigned long)sec_diff < milliseconds / 1000)
-            milliseconds = (unsigned long)sec_diff * 1000;
+    sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    if (sec_diff < 0)
+        return 0; /* clearly timeout */
+
+    /* now sleep at most the given number of milliseconds */
+    if (polling_millisec > 1000) {
+
+        if ((unsigned long)sec_diff < polling_millisec / 1000)
+            polling_millisec = (unsigned int)sec_diff * 1000;
```

Comment:
This does not work right, if polling_mllisec > 1000 
this will wait 1 seconds, until sec_diff == 0, then
since sec_diff < polling_millisec / 1000
set polling_millisec = 0,
then it will skip the wait below and create 100% CPU load
for the last second....

Reply:
I think this is not completely right, this should not be skipped if polling_millisec <= 1000,
since sec_diff can be 0, but you do not sleep(0)

Reply:
No, if `polling_millisec is > 1000` then the function determines `max(polling_millisec, 1000*sec_def)` and waits accordingly, as mentioned in the comment above that code.

Reply:
but in last second of the timeout, 1000*sec_def is null, and then you have a busy loop.
jut for the last second, but I think that is not what you indend.

Reply:
And actually this computes `polling_millisec = MIN(polling_millisec, 1000*sec_def)` 
not MAX. so it can be much smaller than polling_millisec.

Reply:
Oops, I did actually mean: min.

Reply:
If polling_millisec < 1000 the function does not wait zero time even if sec_diff is 0,
but for the intended nap period.

Reply:
Yes, that case is of course more reasonable.
You wouldn't get that sort of problem if you
consider sec_diff = 0 as timeout.

Reply:
Yes, but as documented I'd like to give the connect a last chance in case the remaining time before timeout is less than the resolution of the timeout value, which is one second.